### PR TITLE
K8s postgres triage

### PR DESCRIPTION
**Pull request by stewartshea**
This should also close out https://github.com/runwhen-contrib/rw-public-codecollection/issues/54

I'm leaving it as a draft for now until we can successfully test this in the UI. The local version tests fine but we are hitting OOM errors in the UI. 

I also suspect there will be enough discussion and changes here with @jon-funk  - so the draft is opened to start the review process.

### CODE REVIEW

**Comment by stewartshea** [comment]
- [ ] Note to self to re-read the docs and validate use case configurations once we have the OOM UI issues fixed.

File: codebundles/k8s-postgres-query/runbook.robot [hunk 1]
```diff
@@ -29,8 +29,11 @@ Run Postgres Query And Results to Report
     ...    workload_name=${WORKLOAD_NAME}
     ...    workload_namespace=${WORKLOAD_NAMESPACE}
     ...    workload_container=${WORKLOAD_CONTAINER}
+    ...    kubeconfig=${KUBECONFIG}
+    ...    context=${CONTEXT}
+    ...    target_service=${kubectl}
     ${rsp}=    RW.K8s.Shell
-    ...    cmd=${binary_name} exec ${workload} -- bash -c "${templated_query}" --context ${CONTEXT}
+    ...    cmd=${binary_name} exec ${workload} -- bash -c "echo '${QUERY}' > ${QUERY_FILE_PATH} && ${templated_query}" --context ${CONTEXT}
```
Comment (jon-funk) [inline codebundles/k8s-postgres-query/runbook.robot:36]: Not necessary but a `Upload File To Pod` could help with readability

File: libraries/RW/K8s/namespace_tasks_mixin.py [hunk 2]
```diff
@@ -334,4 +335,176 @@ def count_notready_pods (
...
+        output: str = ""
+        pod_names=self.fetch_pod_names_by_label(namespace, context, kubeconfig, target_service, binary_name, resource_labels)
+        for pod_name in pod_names: 
+          cmd = f"{binary_name} logs {pod_name} --tail={log_lines} --all-containers -n {namespace} --context {context}" 
+          log_output: str = self.shell(
+              cmd=cmd,
+              target_service=target_service,
+              kubeconfig=kubeconfig,
+          )
+          output=f"POD LOGS: {pod_name}\n--------\n{log_output}\n--------\n{output}"
```
Comment (jon-funk) [inline libraries/RW/K8s/namespace_tasks_mixin.py:440]: generating output in a standardized way is tricky. I ran into it a lot when coding "Checks" - we should look at templating output but this is fine for now.

**Review by jon-funk — COMMENTED** [review]
Overall really nice work and some nice new keywords. I have some concerns regarding the pod loops but it should be fine for now.

File: codebundles/k8s-postgres-triage/runbook.robot [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    ...    pattern=\w*
+    ...    default=/tmp/rw-tmp-queries.sql
+    ...    example=/tmp/rw-tmp-queries.sql
+    ${HOSTNAME}=    RW.Core.Import User Variable
+    ...    HOSTNAME
+    ...    type=string
+    ...    description=The hostname specified in the psql connection string. Use localhost, or leave blank, if the execution workload is also hosting the database.
+    ...    pattern=\w*
+    ...    example=localhost
+    ${DISTRIBUTION}=    RW.Core.Import User Variable    DISTRIBUTION
```
[the remark this answers is not in the document]
Comment (stewartshea) [reply codebundles/k8s-postgres-triage/runbook.robot:241]: Agreed, it's also likely the same point to me made about the IF statement above, but I don't really see how to lessen this for the current pattern.

File: codebundles/k8s-postgres-triage/runbook.robot [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    ...    cmd=kubectl get all -l ${RESOURCE_LABELS} -n ${NAMESPACE} --context ${CONTEXT}
+    ...    target_service=${kubectl}
+    ...    kubeconfig=${KUBECONFIG}
+    ${history}=    RW.K8s.Pop Shell History
+    ${history}=    RW.Utils.List To String    data_list=${history}
+    RW.Core.Add Pre To Report    ${stdout}
+    RW.Core.Add Pre To Report    Commands Used: ${history}
+
+Describe Custom Resources
+    IF    "${INCLUDE_CUSTOM_RESOURCES}" == "Yes"
```
[the remark this answers is not in the document]
Comment (stewartshea) [reply codebundles/k8s-postgres-triage/runbook.robot:29]: Moving this to slack.

File: codebundles/k8s-postgres-triage/README.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+  name: [service-account]-sa
+  namespace: [namespace]
+roleRef:
+  kind: Role
+  name: [service-account]-role
+  apiGroup: rbac.authorization.k8s.io
+
+```
+
+## Requirements
```
[the remark this answers is not in the document]
Comment (stewartshea) [reply codebundles/k8s-postgres-triage/README.md:167]: thx!

File: libraries/RW/K8s/namespace_tasks_mixin.py [hunk 6]
```diff
@@ -334,4 +335,176 @@ def count_notready_pods (
...
+        search_filter=f"items[?contains(metadata.name, `{crd_filter}`)].metadata.name"
+        cmd = f"{binary_name} get crd --context {context} -o json" 
+        crd_list: str = self.shell(
+            cmd=cmd,
+            target_service=target_service,
+            kubeconfig=kubeconfig,
+        )
+        # Log search filter - keep this so that useres can validate their patterns with jmespath
+        BuiltIn().run_keyword('Log', search_filter)
+        crd_name_list = search_json(data=json.loads(crd_list), pattern=search_filter)
```
[the remark this answers is not in the document]
Comment (stewartshea) [reply libraries/RW/K8s/namespace_tasks_mixin.py:371]: thx!

File: libraries/RW/K8s/namespace_tasks_mixin.py (hunk 2 above)
Comment (stewartshea) [reply libraries/RW/K8s/namespace_tasks_mixin.py:440]: Agreed... this was very hacky to just get something readable... I'd like a better report template for sure!

File: codebundles/k8s-postgres-query/runbook.robot (hunk 1 above)
Comment (stewartshea) [reply codebundles/k8s-postgres-query/runbook.robot:36]: Are you thinking that as a separate keyword? If so, I was combining it here to save on the amount of API calls made (e.g. avoiding one call to drop in the file and another for the execution), while also trying to keep them in the same bash execution to guarantee that they reside on the same pod. I suppose that latter concern is likely squashed by the fact that the tempalted workload uses a single pod name...

**Review by jon-funk — APPROVED** [review]

**Pull request merged**